prusti-interface/src/environment/body.rs

```diff
@@ -0,0 +1,280 @@
+/// or imported from external crates, all of which are indexed by DefId
+pub struct EnvBody<'tcx> {
+    tcx: TyCtxt<'tcx>,
+
+    local_impure_fns: DynamicallyLoadedBodies<BodyWithBorrowckFacts<'tcx>>,
+    // Loop invariants or triggers
+    local_closures: DynamicallyLoadedBodies<MirBody<'tcx>>,
+
+    pure_fns: PreLoadedBodies<'tcx>,
+    predicates: PreLoadedBodies<'tcx>,
+    specs: PreLoadedBodies<'tcx>,
+
+    /// Copies of above MIR bodies with the given substs applied.
+    monomorphised_bodies: RefCell<FxHashMap<(DefId, SubstsRef<'tcx>), MirBody<'tcx>>>,
+}
+
+impl<'tcx> EnvBody<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>) -> Self {
+        EnvBody {
+            tcx,
+            local_impure_fns: RefCell::new(FxHashMap::default()),
+            local_closures: RefCell::new(FxHashMap::default()),
+            pure_fns: PreLoadedBodies::new(),
+            predicates: PreLoadedBodies::new(),
+            specs: PreLoadedBodies::new(),
+            monomorphised_bodies: RefCell::new(FxHashMap::default()),
+        }
+    }
+
+    // Common logic used internally by functions below to get local MIR bodies.
+    fn raw_load_local_mir(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> BodyWithBorrowckFacts<'tcx> {
+        // SAFETY: This is safe because we are feeding in the same `tcx`
+        // that was used to store the data.
+        let body_with_facts = unsafe { mir_storage::retrieve_mir_body(tcx, def_id) };
+
+        let facts = BorrowckFacts {
+            input_facts: RefCell::new(Some(body_with_facts.input_facts)),
+            output_facts: body_with_facts.output_facts,
+            location_table: RefCell::new(Some(body_with_facts.location_table)),
+        };
+
+        BodyWithBorrowckFacts {
+            body: MirBody(Rc::new(body_with_facts.body)),
+            borrowck_facts: Rc::new(facts),
+        }
+    }
+
+    fn get_monomorphised(&self, def_id: DefId, substs: SubstsRef<'tcx>) -> Option<MirBody<'tcx>> {
+        self.monomorphised_bodies.borrow().get(&(def_id, substs)).cloned()
+    }
+    fn set_monomorphised(
+        &self,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        body: MirBody<'tcx>,
+    ) -> MirBody<'tcx> {
+        if let Entry::Vacant(v) =
+                self.monomorphised_bodies.borrow_mut().entry((def_id, substs)) {
+            v.insert(MirBody(ty::EarlyBinder(body.0).subst(self.tcx, substs))).clone()
+        } else { unreachable!() }
+    }
+
+    /// Get the MIR body of a local impure function, without any substitutions.
+    pub fn get_impure_fn_body(&self, def_id: LocalDefId) -> MirBody<'tcx> {
+        let mut impure = self.local_impure_fns.borrow_mut();
+        impure.entry(def_id).or_insert_with(|| Self::raw_load_local_mir(self.tcx, def_id)).body.clone()
+    }
+
+    /// Get the MIR body of a local impure function, monomorphised
+    /// with the given type substitutions.
+    pub fn get_impure_fn_body_subs(
```

Comment:
`subs` here (and all others) stands for "substitutions"? `subst` might be more consistent. However, I would also expect most of the calls to use the substituting variant, in which case the other ones can be called e.g. `get_*_ident` or `_identity`.